README.md
<<<<<<< HEAD
# Open Cancer TherApeutic Discovery (OCTAD) database package

### Package overview
As the field of precision medicine progresses, we start to tailor treatments for cancer patients classified not only by their clinical, but also by their molecular features. The emerging cancer subtypes defined by these features require dedicated resources to assist the discovery of drug candidates for preclinical evaluation. Voluminous cancer patient gene expression profiles have been accumulated in public databases, enabling the creation of a cancer-specific expression signature. Meanwhile, large-scale gene expression profiles of chemical compounds became available in recent years. By matching the cancer-specific expression signature to compound-induced gene expression profiles from large drug libraries, researchers can prioritize small molecules that present high potency to reverse expression of signature genes for further experimental testing of their efficacy. This approach has proven to be an efficient and cost-effective way to identify efficacious drug candidates. However, the success of this approach requires multiscale procedures, imposing significant challenges to many labs. Therefore, we present OCTAD (http://octad.org): an open workspace for virtually screening compounds targeting precise cancer patient groups using gene expression features. We have included 19,127 patient tissue samples covering 50 cancer types, and expression profiles for 12,442 distinct compounds.  We will continue to include more tissue samples. We streamline all the procedures including deep-learning based reference tissue selection, disease gene expression signature creation, drug reversal potency scoring, and in silico validation. We release OCTAD as a web portal and a standalone R package to allow experimental and computational scientists to easily navigate the tool. The code and data can also be employed to answer various biological questions.

### How to install
To install the package run the following code:
```{r eval=FALSE} 
devtools::install_github('Bin-Chen-Lab/octad.db')
``` 

# Please note, this is a support package for the main package ```octad``` which can be obtained [there](https://github.com/Bin-Chen-Lab/octad)

  
  
###  Contacts and citation
  If you use our work, please cite the [OCTAD paper](https://www.nature.com/articles/s41596-020-00430-z). Both OCTAD package and website was developed by [Bin Chen laboratory](http://binchenlab.org/).
=======
# Web version:
http://octad.org/

# How to Install
Before library installation install required Bioconductor and CRAN packages through this code:
```r
bioconductor_packages=c('edgeR','RUVSeq','DESeq2','limma','rhdf5','artMS')

#For R version 3.5> use BiocManager to install required bioconductor packages: 
if (length(setdiff(bioconductor_packages, rownames(installed.packages()))) > 0) {
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install(setdiff(bioconductor_packages, rownames(installed.packages())))
}

#For R version <3.5 use the BiocInstaller to install required bioconductor packages: 
source("https://bioconductor.org/biocLite.R")
BiocInstaller::biocLite(bioconductor_packages)

packages=c('magrittr','dplyr','ggplot2','doParallel','foreach','lme4','Rfast','httr','data.table')
if (length(setdiff(packages, rownames(installed.packages()))) > 0) {
  install.packages(setdiff(packages, rownames(installed.packages())))  
}
```
Next, install the octad.db, package with all required files for computation available via link  [octad.db](https://chenlab-data-public.s3.amazonaws.com/octad/octad.db_0.99.0.tar.gz%3Fdl%3D0)
```
devtools::install_github('Bin-Chen-Lab/octad.db',build_vignettes = TRUE)
```
It takes a few minutes to install the package and verify files. Afterward, the pipeline will be ready to run. 

```
install.packages("path%to%octad.db_0.99.0.tar.gz", repos = NULL, type="source")
```
Or without downloading the distributive:
```
install.packages("https://chenlab-data-public.s3-us-west-2.amazonaws.com/octad/octad.db_0.99.0.tar.gz",
                 method="libcurl",repos=NULL,type="source")
```
It takes a few minutes to install the package and verify files. Afterward, the pipeline will be ready to run. 
Finally, install the package:
```
devtools::install_github('Bin-Chen-Lab/octad',build_vignettes = TRUE)
```

# Additional data
By default, octad package uses expression data for 978 genes from the LINCS dataset. However, it can influence the result and we advice using whole octad database. To obtatin whole results for DE, downloading of the additional OCTAD database [octad.counts.and.tpm.h5](https://chenlab-data-public.s3-us-west-2.amazonaws.com/octad/octad.counts.and.tpm.h5) from the AWS link is required.

# Tutorial
The tutorial available via following [link](https://chenlab-data-public.s3-us-west-2.amazonaws.com/octad/octad_tutorial.pdf)

# Examples

The several examples listed in the file [octad_example.R](https://github.com/Bin-Chen-Lab/octad_desktop/blob/master/octad_example.R) :

<li>Example 1. liver hepatocellular carcinoma vs adjacent reference tissues;</li> 
<li>Example 2. breast cancer invasive carcinoma with PIK3 mutation vs reference tissues;</li> 
<li>Example 3. lung adenocarcinoma with amplified MYC gene vs reference tissues;</li> 
<li>Example 4. Primary breast cancer invasive carcinoma vs metastatic breast cancer invasive carcinoma using only 978 genes expression data for DE;</li> 
<li>Example 5. Compute sRGES score using GEO obtained dataset</li> 



# Contacts and citation
If you use our work, please cite the paper [OCTAD: an open workplace for virtually screening therapeutics targeting precise cancer patient groups using gene expression features, Nature Protocols](https://www.nature.com/articles/s41596-020-00430-z). Both OCTAD package and website was developed by [Bin Chen laboratory](http://binchenlab.org/).
>>>>>>> origin/master
Examples and questions can be addressed to Eugene Chekalin, PhD, chekali1@msu.edu or Bin Chen, PhD, PI, bin.chen@hc.msu.edu


NEWS
0.99.0
Added a NEWS file to track changes to the package.

R/diffExp.R
#' @export
#' @importFrom    rhdf5 h5read H5close
#' @import RUVSeq edgeR DESeq2 EDASeq stats
#' @importFrom    dplyr everything left_join
#' @importFrom magrittr %>%
#' @importFrom ExperimentHub ExperimentHub

diffExp <- function(case_id='',control_id='',source='octad.small',file='octad.counts.and.tpm.h5',
                                        normalize_samples=TRUE,
                                        k=1,
                    expSet=NULL,
                                        n_topGenes=500,
                                        DE_method='edgeR',
                                        parallel_cores = 2,
                    output=TRUE,
                    outputFolder='', annotate=TRUE){
eh=ExperimentHub()

remLowExpr <- function(counts,counts_phenotype){
    x <-DGEList(counts = round(counts), group = counts_phenotype$sample_type )
    cpm_x <- cpm(x)
    #needs to be at least larger the than the size of the smallest set
    keep.exprs <- rowSums(cpm_x>1) >= min(table(counts_phenotype$sample_type))
    keep.exprs
}



if(missing(case_id)|missing(control_id)){
stop('Case ids and/or control ids vector input not found')
}


if(source=='octad.whole'){

print(paste('loading whole octad expression data for',length(c(case_id,control_id)),'samples',sep=' '))

transcripts = as.character(rhdf5::h5read(file, "meta/transcripts"))
samples = as.character(rhdf5::h5read(file, "meta/samples"))
case_counts = rhdf5::h5read(file, "data/count", index=list(1:length(transcripts), which(samples %in% case_id)))
colnames(case_counts) = samples[samples %in% case_id]
rownames(case_counts) = transcripts
case_id = samples[samples %in% case_id]
normal_counts = rhdf5::h5read(file, "data/count", index=list(1:length(transcripts), which(samples %in% control_id)))
colnames(normal_counts) = samples[samples %in% control_id]
rownames(normal_counts) = transcripts
control_id = samples[samples %in% control_id]
H5close()

expSet = cbind(normal_counts, case_counts)
rm(normal_counts) # free up some memory
rm(case_counts)

}else if(source=='octad.small'){
    print('loading small octad set containing only expression for 978 LINCS genes')
#case.load.start <- Sys.time()
#setwd(base.folder)
#rhdf5_file <- paste0(, "octad.h5")
#transcripts = as.character(rhdf5::h5read(file, "meta/transcripts"))
#samples = as.character(rhdf5::h5read(file, "meta/samples"))
#case_counts = rhdf5::h5read(file, "data/count", index=list(1:length(transcripts), which(samples %in% case_id)))
#colnames(case_counts) = samples[samples %in% case_id]
#rownames(case_counts) = transcripts
#case_id = samples[samples %in% case_id]
#normal_counts = rhdf5::h5read(file, "data/count", index=list(1:length(transcripts), which(samples %in% control_id)))
#colnames(normal_counts) = samples[samples %in% control_id]
#rownames(normal_counts) = transcripts
#control_id = samples[samples %in% control_id]
#H5close()

octad.LINCS.counts=eh[["EH7273"]]
expSet=octad.LINCS.counts[,c(case_id,control_id)]

#expSet=octad.db::octad.LINCS.counts[,c(case_id,control_id)]# bioconductor replace


#expSet = cbind(normal_counts, case_counts)
#rm(normal_counts) # free up some memory
#rm(case_counts)

}else if(source!='octad.small'&source!='octad.small'&missing(expSet)){
stop('Expression data not sourced, please, modify expSet option')

}




    counts_phenotype <- rbind(data.frame(sample = case_id,sample_type = 'case'),
                                                        data.frame(sample = control_id, sample_type = 'control'))
    counts_phenotype = counts_phenotype[counts_phenotype$sample %in% colnames(expSet),]

    counts = expSet[,as.character(counts_phenotype$sample)]
    counts = 2^counts - 1 #unlog the counts it was log(2x + 1) in dz.expr.log2.readCounts
    counts_phenotype$sample = as.character(counts_phenotype$sample)
    counts_phenotype$sample_type = factor(counts_phenotype$sample_type, levels = c("control", "case"))

    #remove lowly expressed transcripts
    highExpGenes <- remLowExpr(counts,counts_phenotype)
    counts = counts[highExpGenes,]

    set <- EDASeq::newSeqExpressionSet(round(counts),
                                                         phenoData = data.frame(counts_phenotype,row.names=counts_phenotype$sample))

    #normalize samples using RUVSeq
    if (normalize_samples == TRUE){
        #compute empirical genes

        design <- stats::model.matrix(~ sample_type, data = pData(set))
        y <- edgeR::DGEList(counts=counts(set), group =    counts_phenotype$sample)
        y <- edgeR::calcNormFactors(y, method="TMM") #upperquartile generate Inf in the LGG case
        y <- edgeR::estimateGLMCommonDisp(y, design)
        y <- edgeR::estimateGLMTagwiseDisp(y, design)
        fit <- edgeR::glmFit(y, design)
        lrt <- edgeR::glmLRT(fit,2) #defaults to compare case control

        top <- edgeR::topTags(lrt, n=nrow(set))$table
        i = which(!(rownames(set) %in% rownames(top)[1:min(n_topGenes,dim(top)[1])]))
        empirical <- rownames(set)[i]
        stopifnot(length(empirical)>0)
        write.csv(data.frame(empirical),file=paste0(outputFolder,"computedEmpGenes.csv"))
        set1 <- RUVSeq::RUVg(set,empirical,k=k)
    }

    if(DE_method=='DESeq2'){
#        library('DESeq2')
        print('computing DE via DESeq')
        row.names(counts_phenotype) = counts_phenotype$sample
        coldata = counts_phenotype

        if (normalize_samples == TRUE){
            dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts(set1),
                                                                        colData = pData(set1),
                                                                        design= ~ sample_type + W_1)
        }else{
            dds <- DESeq2::DESeqDataSetFromMatrix(countData = round(counts),
                                                                        colData = coldata,
                                                                        design= ~ sample_type)
        }
        gc()

        if (parallel_cores > 1){
            dds <- DESeq2::DESeq(dds, parallel = TRUE)
        }else{
            dds <- DESeq2::DESeq(dds)
        }

        #save(dds,file= paste0(outputFolder, "/dds", ".RData"))
        rnms <- DESeq2::resultsNames(dds)

        resRaw <- DESeq2::results(dds, contrast=c("sample_type","case","control"))
        res = data.frame(resRaw)
        res$identifier <- row.names(res)
        res = res %>% select(identifier,everything())
#        return(res)

    }else if(DE_method=='edgeR'){
        print('computing DE via edgeR')

        #construct model matrix based on whether there was normalization ran
        if(normalize_samples == TRUE){
            if(k==1){
                design <- stats::model.matrix(~sample_type + W_1, data=pData(set1))
            }else if(k == 2){
                design <- stats::model.matrix(~sample_type + W_1 + W_2, data = pData(set1))
            }else if (k == 3){
                design <- stats::model.matrix(~sample_type + W_1 + W_2 + W_3, data = pData(set1))
            }
            dgList <- edgeR::DGEList(counts=counts(set1),group=set1$sample_type)

        }else{
            design <- stats::model.matrix(~ sample_type,data=pData(set))
            dgList <- edgeR::DGEList(counts=counts(set),group=set$sample_type)
        }
        dgList <- edgeR::calcNormFactors(dgList, method="TMM") #using upperquartile seems to give issue for LGG
        dgList <- edgeR::estimateGLMCommonDisp(dgList, design)
        dgList <- edgeR::estimateGLMTagwiseDisp(dgList, design)
        fit <- edgeR::glmFit(dgList, design)
        #see edgeRUsersGuide section on testing for DE genes for contrast
        lrt <- edgeR::glmLRT(fit,2)
        #second coefficient otherwise it'll default the W_1 term when normalize is on
        res <- lrt$table
        colnames(res) <- c("log2FoldChange", "logCPM", "LR", "pvalue")
        res$padj <- p.adjust(res$pvalue)
        res$identifier <- row.names(res)
        res = res %>% select(identifier,everything())
#        return(res)
    }else if(DE_method=='limma'){
        #according to https://support.bioconductor.org/p/86461/, LIMMA + VOOM will not use normalized data
        print('computing DE via limma')
#        require('Glimma')
        x <- counts
        nsamples <- ncol(x)
        lcpm <- edgeR::cpm(x, log=TRUE)

        group <-counts_phenotype$sample_type

        ## ----design-----------------------------------------------------------------------------
        design <- model.matrix(~0 + group)
        colnames(design) <- gsub("group", "", colnames(design))

        contr.matrix <- limma::makeContrasts(
            TumorvsNon = case - control,
            levels = colnames(design))

        v <- limma::voom(x, design, plot=FALSE)
        vfit <- limma::lmFit(v, design)
        vfit <- limma::contrasts.fit(vfit, contrasts=contr.matrix)
        efit <- limma::eBayes(vfit)
        #plotSA(efit, main="Final model: Mean variance trend")

        tfit <- limma::treat(vfit, lfc=1) #not_sure
        dt <- limma::decideTests(tfit)
        summary(dt)

        tumorvsnormal <- limma::topTreat(tfit, coef=1, n=Inf)
        tumorvsnormal <- tumorvsnormal[order(abs(tumorvsnormal$logFC), decreasing = TRUE),]
        #tumorvsnormal.topgenes <- rownames(tumorvsnormal[1:50,])


        res <-tumorvsnormal
        colnames(res) <-c("log2FoldChange", "AveExpr", "t", "pvalue", "padj")
        res$identifier = row.names(res)
#        return(res)
    }
#load(merged_gene_info)
if(annotate==TRUE){
	
	merged_gene_info=eh[["EH7272"]]
    #merged_gene_info=octad.db::merged_gene_info #bioconductor replace
	
	
        merged_gene_info$ensembl<-as.vector(merged_gene_info$ensembl)
        merged_gene_info$V1=NULL #modify it when will have time
        merged_gene_info$Symbol=merged_gene_info$gene #modify it when will have time
        merged_gene_info$gene=NULL
        res <- left_join(res, merged_gene_info, by=c('identifier'='ensembl'))
}
return(res)
}
